src/conn/conn_reconfig.c

```diff
@@ -71,40 +76,89 @@ __wt_conn_compat_config(
 			    " for upgrade or downgrade");
 		conn->compat_major = major;
 		conn->compat_minor = minor;
+		F_SET(conn, WT_CONN_COMPATIBILITY);
+	}
+
+	/*
+	 * Only rewrite the turtle file if this is a reconfig. On startup
+	 * it will get written as part of creating the connection.
+	 */
+	if (reconfig) {
+		/*
+		 * We rewrite the turtle file with the metadata information
+		 * that is already there. This call is only to update the
+		 * compatibility information. But since the turtle file is
+		 * entirely rewritten every time we need to send in the
+		 * metadata file information.
+		 */
+		WT_RET(__wt_metadata_search(session, WT_METAFILE_URI, &value));
+		WT_ERR(__wt_metadata_update(session, WT_METAFILE_URI, value));
```

Comment:
I've got two questions about this:

* The ordering seems wrong the following code compares the "saved" version which is retrieved from the metadata - but we've updated the metadata here. Is it comparing the same things? Should we only rewrite the metadata once the versioning checks are done?
* Could we avoid rewriting the metadata if nothing is going to change? That feels nicer than always doing it.

Reply:
The metadata search/update is how we save or remove the compatibility setting from the turtle file. At this point we cannot tell if anything changed because we may be reading this on a restart. Reading the turtle file, parsing and comparing the compatibility contents here, and then conditionally doing the above lines feels like more complicated code for not much gain since we rarely do reconfigures.

Reply:
I do feel uncomfortable re-writing the metadata before doing the checks. Do we have tests that restart to make sure the database isn't left in a weird state after an error is returned? If so I'm happy for this change to be merged.

Reply:
I get it - it's only being re-written here via reconfigure so nothing should happen unless the user has called reconfigure.

I still think it'd be nice not to rewrite the turtle file unless there is an intent to change something in there, but see that it's not as big of a problem.

Reply:
I will see if some reformatting and restructuring helps.

Reply:
@agorrod Thanks for pushing back on this because when I restructured it, I think it uncovered a bug where someone could open with a valid required_min setting and then later reconfigure with a compatibility version that violates that requirement. I'll fix that and make sure the unit test covers that.